Build failure with GHC 8.0.1

I tried to new-build Cabal afresh and I got this error:

```
[ 65 of 203] Compiling Distribution.Types.UnitId ( Distribution/Types/UnitId.hs, /srv/code/cabal-shake/dist-newstyle/build/x86_64-linux/ghc-8.0.1/Cabal-2.1.0.0/noopt/build/Distribution/Types/UnitId.o )

Distribution/Types/UnitId.hs:124:75: error:
    • Couldn't match representation of type ‘forall s2.
                                             P.Parsec
                                               s2 [Distribution.Parsec.Common.PWarning] UnitId’
                               with that of ‘P.Parsec
                                               s1 [Distribution.Parsec.Common.PWarning] DefUnitId’
        arising from a use of ‘ghc-prim-0.5.0.0:GHC.Prim.coerce’
      The data constructor ‘Text.Parsec.Prim.ParsecT’
        of newtype ‘P.ParsecT’ is not in scope
    • In the expression:
          ghc-prim-0.5.0.0:GHC.Prim.coerce
            (parsec ::
               P.Parsec s ([] Distribution.Parsec.Common.PWarning) UnitId) ::
            ParsecParser DefUnitId
      In an equation for ‘parsec’:
          parsec
            = ghc-prim-0.5.0.0:GHC.Prim.coerce
                (parsec ::
                   P.Parsec s ([] Distribution.Parsec.Common.PWarning) UnitId) ::
                ParsecParser DefUnitId
      When typechecking the code for ‘parsec’
        in a derived instance for ‘Parsec DefUnitId’:
        To see the code I am typechecking, use -ddump-deriv
      In the instance declaration for ‘Parsec DefUnitId’

Distribution/Types/UnitId.hs:124:75: error:
    • Couldn't match representation of type ‘forall s2.
                                             P.Parsec
                                               s2 [Distribution.Parsec.Common.PWarning] UnitId’
                               with that of ‘P.Parsec
                                               s1 [Distribution.Parsec.Common.PWarning] DefUnitId’
        arising from a use of ‘ghc-prim-0.5.0.0:GHC.Prim.coerce’
      The data constructor ‘Text.Parsec.Prim.ParsecT’
        of newtype ‘P.ParsecT’ is not in scope
    • In the expression:
          ghc-prim-0.5.0.0:GHC.Prim.coerce
            (lexemeParsec ::
               P.Parsec s ([] Distribution.Parsec.Common.PWarning) UnitId) ::
            ParsecParser DefUnitId
      In an equation for ‘lexemeParsec’:
          lexemeParsec
            = ghc-prim-0.5.0.0:GHC.Prim.coerce
                (lexemeParsec ::
                   P.Parsec s ([] Distribution.Parsec.Common.PWarning) UnitId) ::
                ParsecParser DefUnitId
      When typechecking the code for ‘lexemeParsec’
        in a derived instance for ‘Parsec DefUnitId’:
        To see the code I am typechecking, use -ddump-deriv
      In the instance declaration for ‘Parsec DefUnitId’
```

Oddly, there was no complaint from CI, so I am not sure what is going on...

--- a/Cabal/Distribution/Types/UnitId.hs
+++ b/Cabal/Distribution/Types/UnitId.hs
@@ -121,7 +121,12 @@ getHSLibraryName uid = "HS" ++ display uid
 -- that a 'UnitId' identified this way is definite; i.e., it has no
 -- unfilled holes.
 newtype DefUnitId = DefUnitId { unDefUnitId :: UnitId }
-  deriving (Generic, Read, Show, Eq, Ord, Typeable, Data, Binary, NFData, Parsec, Pretty, Text)
+  deriving (Generic, Read, Show, Eq, Ord, Typeable, Data, Binary, NFData, Pretty, Text)
+
+-- Workaround for a GHC 8.0.1 bug, see
+-- https://github.com/haskell/cabal/issues/4793#issuecomment-334258288
+instance Parsec DefUnitId where
+  parsec = DefUnitId <$> parsec
 
 -- | Unsafely create a 'DefUnitId' from a 'UnitId'.  Your responsibility
 -- is to ensure that the 'DefUnitId' invariant holds.

